### DISCUSSION TITLE
페이지네이션과 무한 스크롤링의 차이를 말해주세요.

### DISCUSSION THREAD

**Discussion by le2sky** [2024-woowacourse-study/level-interview#198] [Database]
뭘까용

**Comment by alstn113** [comment]
페이지네이션은 OLTP 환경에서 사용자에게 반환되는 데이터를 특정 기준으로 정렬된 일부 데이터만을 제공하는 방법론입니다.
페이지네이션에는 크게 두 가지 방식이 있습니다: Offset 기반과 Cursor(Keyset) 기반 페이지네이션입니다.

### Offset 기반 페이지네이션

Offset 기반 페이지네이션은 가장 일반적으로 사용되는 방식으로, 특정 페이지 번호에 해당하는 데이터를 조회하는 방식입니다.
이 방식의 구현 및 쿼리 구성이 간단하며, 첫 페이지 조회 시 성능이 좋습니다.
하지만 페이지가 뒤로 갈수록 읽어야 하는 데이터의 양이 증가하여 성능이 저하되며 많은 데이터를 건너뛸 때 비효율적입니다.

### Cursor(Keyset) 기반 페이지네이션

Cursor 기반 페이지네이션은 무한 스크롤링을 구현할 때 흔히 사용되는 방식으로, 특정 커서 값 이후의 데이터를 조회하는 방식입니다.
이 방식은 항상 일정한 성능을 유지하며, 커서 값이 유일해야 합니다.
구현이 복잡할 수 있으며, 특정 시점 이후의 데이터를 효율적으로 조회할 수 있습니다.
한 번에 여러 페이지를 건너뛸 수 없어 불연속적인 페이지 이동이 어려울 수 있습니다.

### 추가 정보

페이지네이션 성능을 최적화하려면 적절한 인덱스 구성이 필수적입니다.
인덱스가 없으면 성능 개선 효과를 기대할 수 없습니다.

참고: https://one-armed-boy.tistory.com/entry/페이지네이션-방법론-비교Offset-vs-Cursor
성능 개선: https://taegyunwoo.github.io/tech/Tech_DBPagination

**Comment by 3Juhwan** [comment]
페이지네이션과 무한 스크롤링은 웹사이트의 콘텐츠를 나눠서 사용자에게 제공하는 두 가지 방식입니다. 

- 페이지네이션은 콘텐츠를 여러 페이지로 나누어, 사용자가 특정 페이지 번호를 클릭하거나 다음/이전 버튼을 클릭해 다른 페이지로 이동하도록 하는 방식입니다.
- 무한 스크롤링은 사용자가 스크롤을 내릴 때마다 콘텐츠가 자동으로 로드되어 끝없이 이어지는 방식입니다.

</br>

두 방식은 UI/UX 측면에서 다르지만 내부 구현은 유사합니다. 다음은 구현 방법 2가지를 소개합니다. 

**오프셋 기반 페이지네이션**

- 데이터를 특정 페이지 단위로 나누고, 페이지 번호와 각 페이지의 크기를 사용하여 데이터를 가져오는 방식입니다. 
- SQL 쿼리에서 LIMIT와 OFFSET을 사용합니다. 예시) `SELECT * FROM table_name LIMIT 10 OFFSET 20;`
- 데이터 양이 많을 경우, 높은 OFFSET 값에서 성능 저하가 발생할 수 있습니다. N번째 OFFSET을 읽기 위해 이전의 모든 레코드를 읽어야 하기 때문입니다. 
- 페이지의 데이터가 변경되면(예: 삽입, 삭제) 페이지네이션의 결과가 일관되지 않을 수 있습니다.

**커서 기반 페이지네이션**

- 마지막으로 가져온 데이터의 커서를 저장하고, 그 커서 이후의 데이터를 가져오는 방식입니다.
- 주로 고유한 정렬 가능한 필드(예: ID, 타임스탬프)를 사용하여 커서를 설정합니다. 예시) `SELECT * FROM table_name WHERE id > last_id LIMIT 10` 마지막에 가져온 데이터의 last_id부터 10개의 레코드를 조회합니다. 
- 성능이 뛰어나며, 특히 대규모 데이터셋에서 유리합니다. 
- 데이터의 삽입, 삭제에 영향을 덜 받아 결과의 일관성을 유지할 수 있습니다.

**Comment by 3Juhwan** [comment]
오프셋 기반 페이지네이션
- offset & limit
- 높은 offset은 성능 이슈
- 데이터의 삽입, 삭제에 영향을 많이 받음

커서 기반 페이지네이션
- 마지막 데이터의 커서를 저장
- 고유하고 정렬 가능한 필드를 커서로 사용
- 대규모 데이터셋에서 유리
- 데이터의 삽입, 삭제에 일관성 유지
